mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -0,0 +1,428 @@
+
+use super::parquet_prefetch::RangeKey;
+use super::{MetadataRef, RawMetadataRef};
+
+type RowGroupIndex = usize;
+type ColumnIndex = usize;
+
+#[derive(Debug)]
+pub struct ParquetPartStream<Runtime> {
+    runtime: Runtime,
+}
+
+impl<Runtime> ParquetPartStream<Runtime> {
+    pub fn new(runtime: Runtime) -> Self {
+        Self { runtime }
+    }
+}
+
+#[async_trait]
+impl<Runtime> ObjectPartStream for ParquetPartStream<Runtime>
+where
+    Runtime: Spawn,
+{
+    #[allow(clippy::too_many_arguments)]
+    fn spawn_get_object_request<Client>(
+        &self,
+        client: &Client,
+        bucket: &str,
+        key: &str,
+        if_match: ETag,
+        range: RequestRange,
+        preferred_part_size: usize,
+        in_mem_cache: InMemoryCache,
+        parsed_metadata: MetadataRef,
+        raw_metadata: RawMetadataRef,
+    ) -> RequestTask<Client::ClientError>
+    where
+        Client: ObjectClient + Clone + Send + Sync + 'static,
+    {
+        let start = range.start();
+        let size = range.len();
+        let (part_queue, part_queue_producer) = unbounded_part_queue();
+
+        let request_task = {
+            let client = client.clone();
+            let bucket = bucket.to_owned();
+            let id = ObjectId::new(key.to_owned(), if_match);
+            let span = debug_span!("prefetch", range=?range);
+
+            async move {
+                let metadata = parsed_metadata.read().await.clone();
+                let request_range = range.start()..range.end();
+
+                let mut remaining_range = request_range.clone();
+                let mut metadata_range = None;
+                let mut metadata_present = false;
+
+                if let Some(raw_metadata_range) = get_raw_metadata_range(&raw_metadata).await {
+                    if let Some(intersection) = intersect_ranges(&remaining_range, &raw_metadata_range) {
+                        trace!("Metadata portion detected: fetching from cache");
+                        metadata_range = Some(intersection.clone());
+                        metadata_present = true;
+                        remaining_range = remaining_range.start..intersection.start;
+                    }
+                }
+
+                let mut rowgroup_cols: Vec<((RowGroupIndex, ColumnIndex), Range<u64>)> =
+                    if let Some(metadata) = &metadata {
+                        get_row_groups_and_columns(metadata, remaining_range.clone())
+                    } else {
+                        Vec::new()
+                    };
+
+                while !remaining_range.is_empty() {
+                    if !try_serve_from_cache(
+                        &mut remaining_range,
+                        &metadata,
+                        &in_mem_cache,
+                        &id,
+                        &part_queue_producer,
+                        &rowgroup_cols,
+                    )
+                    .await
+                    {
```

Comment:
Put it in its own variable. Same for `fetch_from_client` later.

```suggestion
                    let cache_read_success = try_serve_from_cache(
                        &mut remaining_range,
                        &metadata,
                        &in_mem_cache,
                        &id,
                        &part_queue_producer,
                        &rowgroup_cols,
                    ).await;
                    if !cache_read_success {
```